carp.studies.core/src/commonMain/kotlin/dk/cachet/carp/studies/domain/Study.kt

```diff
@@ -23,14 +25,23 @@ class Study(
     /**
      * A descriptive name for the study, assigned by, and only visible to, the [StudyOwner].
      */
-    var name: String,
+    name: String,
     /**
      * A description of the study, shared with participants once they are invited to the study.
      */
     val invitation: StudyInvitation = StudyInvitation.empty(),
     val id: UUID = UUID.randomUUID()
-)
+) : AggregateRoot<Study, StudySnapshot, Study.Event>()
 {
+    sealed class Event : Immutable()
```

Comment:
@Whathecode Wonder if there should be a `StudyAdded` event as well that simply references itself or is empty. This way we can leave out the repo `add` method as a separate thing and the initial creation of the entity becomes part of the event log (when present in the infrastructure)

Reply:
Isn't that implied? If we remove `add` and only have `store` (or our old `update`), if you then pass an object which does not exist, that you have to add it? But, I'm guessing you are already looking towards full fledged event sourcing, perhaps, in which case we'd need an `added` event either way?

Reply:
My main question would be, who decides when that event should be added? The aggregate root takes care of all of its internals, but not itself. Whether it is initialized through the constructor, or snapshot, there should be no real difference. Then it is the application service that knows when an AR is new. But, opening up raising events breaks encapsulation.I think it is cleaner to have creation separated from

Reply:
hmm...but it's more than the semantics of non-existence. It's also the semantics of dispatching an empty event queue on `store`. We can argue this means it has to be created and error if it exists already. Ok. Let's leave it implicit for now, and then add an explicit event if we see reason to do so.